src/Drupal/Commands/field/FieldDefinitionCommands.php

```diff
@@ -0,0 +1,154 @@
+            'settings' => 'Settings',
+            'class' => 'Class',
+            'provider' => 'Provider',
+        ],
+    )]
+    #[CLI\DefaultFields(fields: ['id', 'label', 'default_settings', 'field_types'])]
+    #[CLI\Help(
+        description: 'Lists field widgets.',
+        synopsis: 'drush field:widgets [options]',
+    )]
+    #[CLI\FilterDefaultField(field: 'id')]
+    public function widgets(array $options = ['format' => 'yaml', 'field-type' => NULL]): RowsOfFields {
+        $processor = static fn(array $definition): array => [
+            'id' => $definition['id'],
+            'label' => (string) $definition['label'],
+            'default_settings' => $definition['class']::defaultSettings(),
+            'field_types' => $definition['field_types'],
+        ];
+        $definitions = \array_map($processor, $this->widgetPluginManager->getDefinitions());
+        if ($options['field-type']) {
+            $definitions = self::filterByFieldType($definitions, $options['field-type']);
+        }
+        return new RowsOfFields(
+            self::encodeDefinitions($definitions, $options['format']),
+        );
+    }
+
+    #[CLI\Command(name: 'field:formatters')]
+    #[CLI\Option(name: 'field-type', description: 'Applicable field type')]
+    #[CLI\FieldLabels(
+        labels: [
+            'id' => 'ID',
+            'label' => 'Label',
+            'default_settings' => 'Default Settings',
+            'field_types' => 'Field types',
+            'settings' => 'Settings',
+            'class' => 'Class',
+            'provider' => 'Provider',
+        ],
+    )]
+    #[CLI\DefaultFields(fields: ['id', 'label', 'default_settings', 'field_types'])]
+    #[CLI\Help(
+        description: 'Lists field formatters.',
+        synopsis: 'drush field:formatters [options]',
+    )]
+    #[CLI\FilterDefaultField(field: 'id')]
+    public function formatters(array $options = ['format' => 'yaml', 'field-type' => NULL]): RowsOfFields {
+        $processor = static fn(array $definition): array => [
+            'id' => $definition['id'],
+            'label' => (string) $definition['label'],
+            'default_settings' => $definition['class']::defaultSettings(),
+            'field_types' => $definition['field_types'],
+            'class' => $definition['class'],
+            'provider' => $definition['provider'],
+        ];
+        $definitions = \array_map($processor, $this->formatterPluginManager->getDefinitions());
+        if ($options['field-type']) {
+            $definitions = self::filterByFieldType($definitions, $options['field-type']);
+        }
+        return new RowsOfFields(
+            self::encodeDefinitions($definitions, $options['format']),
+        );
+    }
+
+    /**
+     * Encodes rows.
+     *
+     * Some output formats i.e. 'table' expect that each row to be a scalar
+     * value.
+     */
+    private static function encodeDefinitions(array $definitions, string $format): mixed {
```

Comment:
See https://github.com/drush-ops/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/StatusCommands.php#L87
https://github.com/weitzman/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/StatusCommands.php#L87 for an approach we use in a few places for formatting table cells.

Reply:
Or, do what https://github.com/drush-ops/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/SiteCommands.php#L99
https://github.com/weitzman/drush/blob/d013acfdb3fa0e48282e19640b065d584f9cb12e/src/Commands/core/SiteCommands.php#L99 and others do and default to `format=yaml`. That handles nested data easily.

Reply:
That encoder is needed not only for table format but for some other formats that does not support nested structures. Seems like renderer function is not used for `string` format.

Reply:
I made a PR into this PR with my proposed changes https://github.com/Chi-teck/drush/pull/1